### PR TITLE
Strictly match `next/dist` in externals configuration

### LINKED ISSUE #9022: Invalid hook call in 9.0.6
# Bug report

## Describe the bug

When you use react a component residing outside the main Next.js project folder which uses hooks. You end up getting `Invalid hook call` error and the application breaks. Components without hooks work as expected.

The bug appears in all versions `>9.0.5` when you change the webpack config so that files outside the main folder are transpiled. It’s working fine in `<=9.0.5` 

## To Reproduce

Check out the repro at https://github.com/baldurh/next-9.0.6-bug-repro

## Expected behavior

The code should not break when using files outside the project folder.

## System information

- OS: N/A
- Browser: N/A
- Version of Next.js: `>=9.0.6`

## Additional context

I know this is probably not a common use of Next.js but in our project we’re using a monorepo and have a shared folder with components used by multiple applications. It would be nice to get this working again. If someone finds an alternative config we could use I’d also be happy to do that 🙂

### PR DESCRIPTION
**Pull request by felixmosh**
such as `i18next`, `next-i18next`, `react-i18next` to be external
image
fixes #9022

### CODE REVIEW

**Review by timneutkens — APPROVED** [review]

File: packages/next/build/webpack-config.ts [hunk 1]
```diff
@@ -476,7 +476,7 @@ export default async function getBaseWebpackConfig(
             // Default pages have to be transpiled
             if (
               !res.match(/next[/\\]dist[/\\]next-server[/\\]/) &&
```
Comment (timneutkens) [inline packages/next/build/webpack-config.ts]:
```suggestion
              !res.match(/[/\\]next[/\\]dist[/\\]next-server[/\\]/) &&
```

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| buildDuration | 13.5s | 13.6s | ⚠️ +143ms |
| nodeModulesSize | 48.9 MB | 48.9 MB | ⚠️ +10 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.13 kB | 5.13 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| 4952ddcd88e7..54d3.js gzip | 4.68 kB | 4.68 kB | ✓ |
| commons.HASH.js gzip | 4.06 kB | 4.06 kB | ✓ |
| de003c3a9d30..4cf7.js gzip | 13.7 kB | 13.7 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 67.5 kB | 67.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.19 kB | 4.19 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| 4952ddcd88e7..dule.js gzip | 5.56 kB | 5.56 kB | ✓ |
| de003c3a9d30..dule.js gzip | 12.5 kB | 12.5 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 62.1 kB | 62.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 4.76 kB | 4.76 kB | ✓ |
| Overall change | 4.76 kB | 4.76 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _app.js gzip | 1.33 kB | 1.33 kB | ✓ |
| _error.js gzip | 4.07 kB | 4.07 kB | ✓ |
| hooks.js gzip | 779 B | 779 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.9 kB | 2.9 kB | ✓ |
| routerDirect.js gzip | 283 B | 283 B | ✓ |
| withRouter.js gzip | 282 B | 282 B | ✓ |
| Overall change | 9.87 kB | 9.87 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _app.module.js gzip | 757 B | 757 B | ✓ |
| _error.module.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks.module.js gzip | 371 B | 371 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 2.47 kB | 2.47 kB | ✓ |
| routerDirect..dule.js gzip | 273 B | 273 B | ✓ |
| withRouter.m..dule.js gzip | 272 B | 272 B | ✓ |
| Overall change | 7.41 kB | 7.41 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| index.html gzip | 1.02 kB | 1.02 kB | ✓ |
| link.html gzip | 1.03 kB | 1.03 kB | ✓ |
| withRouter.html gzip | 1.02 kB | 1.02 kB | ✓ |
| Overall change | 3.07 kB | 3.07 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| buildDuration | 13.9s | 14s | ⚠️ +141ms |
| nodeModulesSize | 48.9 MB | 48.9 MB | ⚠️ +10 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.13 kB | 5.13 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| 4952ddcd88e7..54d3.js gzip | 4.68 kB | 4.68 kB | ✓ |
| commons.HASH.js gzip | 4.06 kB | 4.06 kB | ✓ |
| de003c3a9d30..4cf7.js gzip | 13.7 kB | 13.7 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 67.5 kB | 67.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.19 kB | 4.19 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| 4952ddcd88e7..dule.js gzip | 5.56 kB | 5.56 kB | ✓ |
| de003c3a9d30..dule.js gzip | 12.5 kB | 12.5 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 62.1 kB | 62.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 4.76 kB | 4.76 kB | ✓ |
| Overall change | 4.76 kB | 4.76 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _app.js gzip | 1.33 kB | 1.33 kB | ✓ |
| _error.js gzip | 4.07 kB | 4.07 kB | ✓ |
| hooks.js gzip | 779 B | 779 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.9 kB | 2.9 kB | ✓ |
| routerDirect.js gzip | 283 B | 283 B | ✓ |
| withRouter.js gzip | 282 B | 282 B | ✓ |
| Overall change | 9.87 kB | 9.87 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _app.module.js gzip | 757 B | 757 B | ✓ |
| _error.module.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks.module.js gzip | 371 B | 371 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 2.47 kB | 2.47 kB | ✓ |
| routerDirect..dule.js gzip | 273 B | 273 B | ✓ |
| withRouter.m..dule.js gzip | 272 B | 272 B | ✓ |
| Overall change | 7.41 kB | 7.41 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _error.js gzip | 77.8 kB | 77.8 kB | ✓ |
| hooks.html gzip | 1.06 kB | 1.06 kB | ✓ |
| index.js gzip | 78 kB | 78 kB | ✓ |
| link.js gzip | 80.4 kB | 80.4 kB | ✓ |
| routerDirect.js gzip | 78.1 kB | 78.1 kB | ✓ |
| withRouter.js gzip | 78.1 kB | 78.1 kB | ✓ |
| Overall change | 393 kB | 393 kB | ✓ |

</details>

</details>


Commit: 7289e1f0abce84de07fe89dbc230a4e830cde5ab

**Comment by timneutkens** [comment]
I did some investigating on #9022 and this does not fix the case that is described in the initial issue as that one is related to how React is resolved. However I'd argue that the case described in the initial issue is a wrong project structure (deps not available) and that what you've fixed here is more important 👍

File: packages/next/build/webpack-config.ts (hunk 1 above)
Comment (felixmosh) [reply packages/next/build/webpack-config.ts]: This looks a folder of `next-server` as well so it not applies to other libs :]

**Comment by felixmosh** [comment]
I had the same error as #9022 when updating to ^9.06, as I've wrote in the comment https://github.com/zeit/next.js/issues/9022#issuecomment-570921527,  it is an issue of combination of 2 things, usage of sym-link & the fact that these libs had multiple instance in side server bundle...

I've tested, if one of the conditions is not apply, the bug is not reproduces.
When I've fixed the `next` regex, even sym-link started to work...

**Comment by timneutkens** [comment]
Could you add a test for this? Then we can ensure it keeps working.

**Comment by felixmosh** [comment]
Sure, any pointer of where?

**Comment by baldurh** [comment]
I can confirm this PR does not fix #9022 in my case. See my https://github.com/zeit/next.js/issues/9022#issuecomment-571152664.

@timneutkens I wouldn’t say this is a result of a _wrong_ project structure, but rather an unusual project structure. However, I’m not the only one as can be seen in the original issue.

**Comment by felixmosh** [comment]
> I can confirm this PR does not fix #9022 in my case. See my [#9022 (comment)](https://github.com/zeit/next.js/issues/9022#issuecomment-571152664).
> 
> @timneutkens I wouldn’t say this is a result of a _wrong_ project structure, but rather an unusual project structure. However, I’m not the only one as can be seen in the original issue.

I've double checked, it solves the issue that I had.
I didn't had weird structure, but a mono-repo with usage of sym-link for building the production build & usage of `i18next` that got bundled in this case inside the server bundles.

**Comment by timneutkens** [comment]
@felixmosh your case is fine and this PR will definitely fix it. Baldur's case is that there is no node_modules folder outside of the subdirectory and that causes React to not be resolved. Then an alias was added to next.config.js but that doesn't work as expected. Particularly what changed that caused that issue in particular is yarn PnP support.

Note that if you have a Node.js project and you would `require` the file it would break in exactly the same way, hence why I'm saying the project structure is wrong.

**Comment by baldurh** [comment]
@timneutkens good point about Node.js. I guess you are right 😅

**Review by Timer — APPROVED** [review]
I'll handle the tests, thanks!

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| buildDuration | 13.1s | 13.1s | ⚠️ +51ms |
| nodeModulesSize | 48.9 MB | 48.9 MB | ⚠️ +10 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.13 kB | 5.13 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| 4952ddcd88e7..54d3.js gzip | 4.68 kB | 4.68 kB | ✓ |
| commons.HASH.js gzip | 4.06 kB | 4.06 kB | ✓ |
| de003c3a9d30..4cf7.js gzip | 13.7 kB | 13.7 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 67.5 kB | 67.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.19 kB | 4.19 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| 4952ddcd88e7..dule.js gzip | 5.56 kB | 5.56 kB | ✓ |
| de003c3a9d30..dule.js gzip | 12.5 kB | 12.5 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 62.1 kB | 62.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 4.76 kB | 4.76 kB | ✓ |
| Overall change | 4.76 kB | 4.76 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _app.js gzip | 1.33 kB | 1.33 kB | ✓ |
| _error.js gzip | 4.07 kB | 4.07 kB | ✓ |
| hooks.js gzip | 779 B | 779 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.9 kB | 2.9 kB | ✓ |
| routerDirect.js gzip | 283 B | 283 B | ✓ |
| withRouter.js gzip | 282 B | 282 B | ✓ |
| Overall change | 9.87 kB | 9.87 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _app.module.js gzip | 757 B | 757 B | ✓ |
| _error.module.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks.module.js gzip | 371 B | 371 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 2.47 kB | 2.47 kB | ✓ |
| routerDirect..dule.js gzip | 273 B | 273 B | ✓ |
| withRouter.m..dule.js gzip | 272 B | 272 B | ✓ |
| Overall change | 7.41 kB | 7.41 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| index.html gzip | 1.02 kB | 1.02 kB | ✓ |
| link.html gzip | 1.03 kB | 1.03 kB | ✓ |
| withRouter.html gzip | 1.02 kB | 1.02 kB | ✓ |
| Overall change | 3.07 kB | 3.07 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| buildDuration | 13.7s | 13.6s | -42ms |
| nodeModulesSize | 48.9 MB | 48.9 MB | ⚠️ +10 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| main-HASH.js gzip | 5.13 kB | 5.13 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| 4952ddcd88e7..54d3.js gzip | 4.68 kB | 4.68 kB | ✓ |
| commons.HASH.js gzip | 4.06 kB | 4.06 kB | ✓ |
| de003c3a9d30..4cf7.js gzip | 13.7 kB | 13.7 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 67.5 kB | 67.5 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 4.19 kB | 4.19 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| 4952ddcd88e7..dule.js gzip | 5.56 kB | 5.56 kB | ✓ |
| de003c3a9d30..dule.js gzip | 12.5 kB | 12.5 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 62.1 kB | 62.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 4.76 kB | 4.76 kB | ✓ |
| Overall change | 4.76 kB | 4.76 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _app.js gzip | 1.33 kB | 1.33 kB | ✓ |
| _error.js gzip | 4.07 kB | 4.07 kB | ✓ |
| hooks.js gzip | 779 B | 779 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.9 kB | 2.9 kB | ✓ |
| routerDirect.js gzip | 283 B | 283 B | ✓ |
| withRouter.js gzip | 282 B | 282 B | ✓ |
| Overall change | 9.87 kB | 9.87 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _app.module.js gzip | 757 B | 757 B | ✓ |
| _error.module.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks.module.js gzip | 371 B | 371 B | ✓ |
| index.module.js gzip | 212 B | 212 B | ✓ |
| link.module.js gzip | 2.47 kB | 2.47 kB | ✓ |
| routerDirect..dule.js gzip | 273 B | 273 B | ✓ |
| withRouter.m..dule.js gzip | 272 B | 272 B | ✓ |
| Overall change | 7.41 kB | 7.41 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | zeit/next.js canary  | felixmosh/next.js fix-9022 | Change |
| - | - | - | - |
| _error.js gzip | 77.8 kB | 77.8 kB | ✓ |
| hooks.html gzip | 1.05 kB | 1.05 kB | ✓ |
| index.js gzip | 78 kB | 78 kB | ✓ |
| link.js gzip | 80.4 kB | 80.4 kB | ✓ |
| routerDirect.js gzip | 78.1 kB | 78.1 kB | ✓ |
| withRouter.js gzip | 78.1 kB | 78.1 kB | ✓ |
| Overall change | 393 kB | 393 kB | ✓ |

</details>

</details>


Commit: ae44d77d1ff884499af8e5b86c0f8cb15334d6a0

**Comment by Timer** [comment]
Test added in https://github.com/zeit/next.js/pull/9968

**Comment by Timer** [comment]
Released as `9.1.8-canary.3`.

**Comment by m-spyratos** [comment]
@tim-phillips I'm watching #9022 because I've setup my project following the sample [NextJS with Firebase Hosting](https://github.com/firebase/functions-samples/tree/master/nextjs-with-firebase-hosting). As soon as I install React Bootstrap or React Material I get this error cause I end up with multiple react instances.

I would suggest to keep #9022 open or at least update the documentation with a better solution. It's a very common scenario libraries like the above to be used alongside React and the proposed Firebase file structure doesn't allow this.

**Pull request merged**